core/server/proxy/src/main/java/alluxio/proxy/s3/signature/StringToSignProducer.java

```diff
@@ -0,0 +1,326 @@
+    StringJoiner result = new StringJoiner(regex);
+    Pattern p = Pattern.compile(regex);
+    Matcher m = p.matcher(uri);
+    int pos = 0;
+    while (m.find()) {
+      result.add(urlEncode(uri.substring(pos, m.start())));
+      pos = m.end();
+    }
+    result.add(urlEncode(uri.substring(pos)));
+    return result.toString();
+  }
+
+  /**
+   * Encode url.
+   *
+   * Spaces must be encoded as %20. Do not encode spaces as plus signs (+).
+   * Apply the encoding algorithm and replace the plus sign (+) in the encoded
+   * string with %20, the asterisk (*) with %2A, and %7E with the tilde (~).
+   *
+   * @param str string to be encoded
+   * @return the encoded String
+   */
+  private static String urlEncode(String str) {
+    try {
+      return URLEncoder.encode(str, "UTF-8")
+              .replaceAll("\\+", "%20")
+              .replaceAll("%7E", "~");
+    } catch (UnsupportedEncodingException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static String getQueryParamString(Map<String, String> queryMap) {
+    List<String> params = new ArrayList<>(queryMap.keySet());
+
+    // Sort by name, then by value
+    Collections.sort(params, (o1, o2) -> o1.equals(o2)
+            ? queryMap.get(o1).compareTo(queryMap.get(o2))
+            : o1.compareTo(o2));
+
+    StringBuilder result = new StringBuilder();
+    for (String p : params) {
+      if (!p.equals("X-Amz-Signature")) {
+        if (result.length() > 0) {
+          result.append("&");
+        }
+        result.append(urlEncode(p));
+        result.append('=');
+        result.append(urlEncode(queryMap.get(p)));
+      }
+    }
+    return result.toString();
+  }
+
+  static void validateSignedHeader(
+      String schema,
+      String header,
+      String headerValue
+  ) throws S3Exception {
+    switch (header) {
+      case HOST:
+        try {
+          URI hostUri = new URI(String.format("%s://%s", schema, headerValue));
+          InetAddress.getByName(hostUri.getHost());
+        } catch (UnknownHostException | URISyntaxException e) {
+          LOG.error("Host value mentioned in signed header is not valid. "
+                  + "Host:{}", headerValue);
+          throw new S3Exception(headerValue, S3ErrorCode.AUTHINFO_CREATION_ERROR);
+        }
+        break;
+      case X_AMZ_DATE:
+        LocalDate date = LocalDate.parse(headerValue, TIME_FORMATTER);
+        LocalDate now = LocalDate.now();
+        if (date.isBefore(now.minus(PRESIGN_URL_MAX_EXPIRATION_SECONDS, ChronoUnit.SECONDS))
+             || date.isAfter(now.plus(PRESIGN_URL_MAX_EXPIRATION_SECONDS, ChronoUnit.SECONDS))) {
+          LOG.error("AWS date not in valid range. Request timestamp:{} should "
+                 + "not be older than {} seconds.", headerValue,
+                 PRESIGN_URL_MAX_EXPIRATION_SECONDS);
+          throw new S3Exception(headerValue, S3ErrorCode.AUTHINFO_CREATION_ERROR);
+        }
+        break;
+      case X_AMZ_CONTENT_SHA256:
+        break;
```

Comment:
Is there supposed to be a checksum verification here? Otherwise leave a comment explaining why this is empty here.

Reply:
Deleted.